src/mbgl/shader/circle_vertex.cpp

```diff
@@ -0,0 +1,13 @@
+#include <mbgl/shader/circle_vertex.hpp>
+#include <mbgl/shader/shader.hpp>
+#include <mbgl/gl/gl.hpp>
+
+namespace mbgl {
+
+void CircleVertex::bind(const int8_t* offset) {
+    static_assert(sizeof(CircleVertex) == 4, "expected CircleVertex size");
+
+    MBGL_BIND_VERTEX_ATTRIBUTE(CircleVertex, a_pos, offset);
```

Comment:
Vertex attribute binding involves some shader-specific things and some vertex-specific things:

* Shader-specific: attribute name, index, and GLSL type.
* Vertex-specific: member name, offset, and C++ type.

I'd like to come up with a design where:

* A given C++ vertex type can be used with multiple shaders.
* A shader requires one particular vertex type, enforced at compile time.
* Attribute indexes are defined per-`Shader`-subclass, rather than in `Shader`, as they are now. Not all shaders have all attributes.
* Debug builds verify that the GLSL and native types are consistent, using `glGetActiveAttrib` to introspect the GLSL type. (We should do this for uniforms as well.)
* I'm open to a design that doesn't enforce that the GLSL attribute name matches the C++ member name, but it also doesn't seem bad to do so. Why would we need them to differ?

I think the starting point here is to introduce a counterpart to `uniform.{hpp,cpp}` for attributes and add members to each `Shader` subclass -- e.g.:

```
AttributeVector<int16_t, 2> a_pos = { "a_pos", *this };
```

Then see where that leads the binding code.